### ISSUE TITLE
Add RPC GetUtxoByAddress

### ISSUE THREAD

**Issue by braddmiller** [zcash/lightwalletd#312]
As a wallet developer I need to be able to retrieve all UTXOs related to a t-address.

Notes:
- This could require enabling the "spentindex" on zcashd

Requirements:
- [ ] List of UTXOs

**Comment by LarryRuane** [comment]
I verified that it does need `spentindex` (only present when full Insight Explorer option is enabled). We could actually create a lighter-weight index that stores less information than the current `spentindex`; here's the data that index currently stores per transparent output:
```
struct CSpentIndexValue {
    uint256 txid;
    unsigned int inputIndex;
    int blockHeight;
    CAmount satoshis;
    int addressType;
    uint160 addressHash;
}
```
We'd really only need to store the key (txid and output index), and no data, to know if this output is unspent.

**Comment by LarryRuane** [comment]
@pacu, @kevin, how do you want this to work exactly?
- **Argument**
  - One address only or multiple? (I'm guessing one.)
  - Include a block range? Or always just return all?
- **Result**
  - Include the mempool? Or only UTXOs in confirmed blocks?
  - What format is best? Just (txid, index) pairs? 
  - Do you need the actual UTXO (data)? Or can you construct spending transactions without that (and without having to fetch the transactions containing the UTXOs)?
  - Streaming output okay? (There could be a large number of items)

**Comment by LarryRuane** [comment]
Correction, we do not need to enable `spentindex`. That index allows you to determine _which (later) input_ has spent a given output; that is, it's a list of _non_-UTXOs (outputs that _have_ been spent). It's a forward-in-time reference, which the blockchain doesn't provide directly.

There's another index called `addressUnspent` that _is_ present in the `-lightwalletd` configuration. This index maps an address to the set of outputs to the address that are currently unspent.

Here's a real mainnet example of using the zcash rpc to access this index (this particular address's list is actually much longer, showing just 2 here):
```
$ src/zcash-cli getaddressutxos '"t3R3Y5vnBLrEn8L6wFjPjBLnxSUQsKnmFpv"'
[
  {
    "address": "t3R3Y5vnBLrEn8L6wFjPjBLnxSUQsKnmFpv",
    "txid": "9cfb7d65723c4014438ba7ea75b67bc0e90319dd4f75d871e5fefa09fdc687c7",
    "outputIndex": 1,
    "script": "a91447185bab99466b849d774faa2edfa983bd37741487",
    "satoshis": 125000000,
    "height": 1005710
  },
  {
    "address": "t3R3Y5vnBLrEn8L6wFjPjBLnxSUQsKnmFpv",
    "txid": "364bf5c8fc8e9e5ac7b5dd8fcc0cd7102950a73bcdb682af645837a654fb9afe",
    "outputIndex": 1,
    "script": "a91447185bab99466b849d774faa2edfa983bd37741487",
    "satoshis": 125000000,
    "height": 1005711
  }
]
$
```
One small zcashd change is needed. Currently, this `getaddressutxos` zcash rpc is disabled unless `-insightexplorer` is enabled:
```
    if (!fExperimentalInsightExplorer) {
        throw JSONRPCError(RPC_MISC_ERROR, "Error: getaddressutxos is disabled. "
            "Run './zcash-cli help getaddressutxos' for instructions on how to enable this feature.");
    }
```
But the index is there, so all that's needed is to remove this restriction. Existing `-lightwalletd` nodes won't have to reindex.

**Comment by gmale** [comment]
@LarryRuane unfortunately, our answers almost entirely depend on what's needed in librustzcash. So it might be better for @str4d or @nuttycom to weigh in on the data that we need from lightwalletd. Here are my initial thoughts:

**Argument**
- One address
- Include a block range
**Result**
- Only UTXOs in confirmed blocks
- What format is best? not sure
- Do you need the actual UTXO (data)? not sure
- Streaming output okay? I tend to not use streaming

**Comment by LarryRuane** [comment]
@gmale, you wrote:
> Include a block range

Would your answer be different if you knew that this creates no efficiency saving (other than the obvious one of sending you less data)? The zcash `getaddressutxos` rpc doesn't have a block-range filter; it always sends all of them. I can then filter what I send to you if you're sure you want a block range. But I was trying to think why the range restriction would be useful. I'd think any and all UTXOs are fair game. If we're doing autoshielding, there probably won't typically be very many UTXOs (other than perhaps just after importing a wallet) so would sending all be okay?

We'd still like to have @str4d or @nuttycom advise us on what information this gRPC would return to the wallet.

**Comment by gmale** [comment]
I was just thinking in terms of how the scan might work. Having a range would not prevent us from scanning everything but not having a range would prevent us from scanning from a known height. Can the range be optional?

**Comment by LarryRuane** [comment]
I see clearly now, I think; the advantage of the range part of the argument is that you might get all of them initially and cache them, and later ask for UTXOs starting at one beyond the last height you requested at previously. That's much better than having to receive the entire set every time (which makes caching impossible).

(Side note: If you do cache UTXOs, it's possible that what you think is a UTXO becomes a _TXO_ outside your knowledge or control, in other words, the cache could become stale, but that seems unlikely. Maybe just document how and why this can happen, and give the user a button to push to rescan for UTXOs, i.e., rebuild the cache from scratch.)

If my understanding is correct, then a better interface (than a range) would be just the starting height, and I'd return everything at that height and higher, up to the current tip. You could do incremental updating of your UTXO cache.

We could also have a _max number_ of UTXOs to return, if there's a limit to how many you want to receive in one reply. I think this is better than the range approach because you might ask for a range that contains no entries at all, which is wasteful. With a start height and number of entries (each UTXO entry returns the height of the block that contains it), you could submit a sequence of requests, each beginning with the height of the last entry you received. (The entries are sorted by block height, by the way.) If you received fewer than you requested, then you'd know that there were no more.

To summarize, I think the argument should be either:
- start_height, or
- start_height, max_number_entries

**Comment by LarryRuane** [comment]
I just noticed that this ticket is ambiguous; the requirement here (from the first comment) is:
> List of UTXOs

I'd been assuming this meant replying with the kinds of information that is returned by `getaddressutxos` (see the [comment](https://github.com/zcash/lightwalletd/issues/312#issuecomment-708111218) above).

But all you will have is the _reference_  to the UTXO (its transaction ID and the output index), and its amount, but not the UTXO itself. Is that enough to create the spending transaction? Or will the wallet still need to fetch the transaction (using the existing `GetTransaction` gRPC) to access the actual UTXO (data) itself?

If the latter, should this new gRPC do that transaction fetching and reply with the UTXO (data), in addition to the reference (txid and index) so the wallet doesn't have to do the `GetTransaction`? This would require the lightwalletd to do a separate zcash rpc call (to get the transaction), but that's very efficient (because lightwalletd is co-located with the zcashd), much more efficient than requiring the wallet to separately fetch the transactions.
